### PR TITLE
[AMDGPU] Sign extend simm16 in setreg intrinsic

### PR DESCRIPTION
**Pull request by rampitec**
We currently force users to use a negative contant in the intrinsic call. Changing it zext would break existing programs, so just sign extend an argument.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SOPInstructions.td [hunk 1]
```diff
@@ -1160,6 +1158,9 @@ def S_SETREG_IMM32_B32_mode : S_SETREG_IMM32_B32_Pseudo {
 
 } // End Defs = [MODE], Uses = [MODE]
 
+def : GCNPat<(int_amdgcn_s_setreg (i32 timm:$simm16), i32:$sdst),
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SOPInstructions.td:1161]: Could define an SDNodeXForm for TimmLeaf and avoid splitting out the pattern from the instruction definition

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SOPInstructions.td]: Done

**Review by arsenm — APPROVED** [review]

**Comment by fmayer** [comment]
This broke UBSan bots

```
RUN: at line 2: /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -march=amdgcn -mcpu=verde -verify-machineinstrs -show-mc-encoding < /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll | /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck -check-prefixes=GFX6 /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -march=amdgcn -mcpu=verde -verify-machineinstrs -show-mc-encoding
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck -check-prefixes=GFX6 /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll
/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/Target/AMDGPU/SIModeRegister.cpp:288:27: runtime error: shift exponent 32 is too large for 32-bit type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/Target/AMDGPU/SIModeRegister.cpp:288:27 in 
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck -check-prefixes=GFX6 /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll
```

**Comment by rampitec** [comment]
> This broke UBSan bots
> 
> ```
> RUN: at line 2: /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -march=amdgcn -mcpu=verde -verify-machineinstrs -show-mc-encoding < /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll | /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck -check-prefixes=GFX6 /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll
> + /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -march=amdgcn -mcpu=verde -verify-machineinstrs -show-mc-encoding
> + /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck -check-prefixes=GFX6 /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll
> /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/Target/AMDGPU/SIModeRegister.cpp:288:27: runtime error: shift exponent 32 is too large for 32-bit type 'int'
> SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/Target/AMDGPU/SIModeRegister.cpp:288:27 in 
> FileCheck error: '<stdin>' is empty.
> FileCheck command line:  /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck -check-prefixes=GFX6 /b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.setreg.ll
> ```

Thanks. The culprit is not the patch itself, but the added test has uncovered pre-existing problem:

```
 unsigned Mask = ((1 << Width) - 1) << Offset;
```
The test runs this part of code with Width == 32 which overflows unsigned int.

**Comment by rampitec** [comment]
> Thanks. The culprit is not the patch itself, but the added test has uncovered pre-existing problem:
> 
> ```
>  unsigned Mask = ((1 << Width) - 1) << Offset;
> ```
> 
> The test runs this part of code with Width == 32 which overflows unsigned int.

https://github.com/llvm/llvm-project/pull/78492

**Pull request merged**